File: GameLibrary/src/cc/lib/swing/AWTSoundMgr.java
package cc.lib.swing;

import java.io.File;
import java.util.ArrayList;

import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioInputStream;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.Clip;
import javax.sound.sampled.DataLine;

public class AWTSoundMgr {

	private AWTSoundMgr() {}
	
    /**
     * 
     * @param file_name
     * @return
     */
    public static int loadAudio(String file_name)
    {
        try {
            // From file
            AudioInputStream stream = AudioSystem.getAudioInputStream(new File(file_name));
        
            // At present, ALAW and ULAW encodings must be converted
            // to PCM_SIGNED before it can be played
            AudioFormat format = stream.getFormat();
            if (format.getEncoding() != AudioFormat.Encoding.PCM_SIGNED) {
                format = new AudioFormat(
                        AudioFormat.Encoding.PCM_SIGNED,
                        format.getSampleRate(),
                        format.getSampleSizeInBits()*2,
                        format.getChannels(),
                        format.getFrameSize()*2,
                        format.getFrameRate(),
                        true);        // big endian
                stream = AudioSystem.getAudioInputStream(format, stream);
            }
        
            // Create the clip
            DataLine.Info info = new DataLine.Info(
                Clip.class, stream.getFormat(), ((int)stream.getFrameLength()*format.getFrameSize()));
            Clip clip = (Clip) AudioSystem.getLine(info);
        
            // This method does not return until the audio file is completely loaded
            clip.open(stream);
        
            sounds.add(clip);
            return sounds.size()-1;
        } catch (Exception e) {
        	System.err.println("EXCEPTION " + e + " caught loading sounds " + file_name);
        } 
        
        return -1;
    }
    
    /**
     * 
     * @param id
     */
    public static void playSound(int id) {
    	if (id<0)
    		return;
    	Clip clip = (Clip)sounds.get(id);
    	if (clip != null)
    		clip.start();
    }
    
    private static ArrayList<Clip> sounds = new ArrayList<Clip>(32);
}

File: SwingApps/src/cc/lib/swing/AWTSoundMgr.kt
package cc.lib.swing

import cc.lib.logger.LoggerFactory
import cc.lib.utils.getOrNull
import java.io.File
import javax.sound.sampled.AudioFormat
import javax.sound.sampled.AudioSystem
import javax.sound.sampled.Clip
import javax.sound.sampled.DataLine

object AWTSoundMgr {

	val log = LoggerFactory.getLoggerForName("AWTSoundMgr")

	private val searchPaths = mutableListOf<File>().also {
		it.add(File("."))
	}

	fun addSearchPath(path: String) {
		File(path).also {
			if (!it.isDirectory)
				throw Exception("Unknown directory $${it.absolutePath}")
			searchPaths.add(it)
		}
	}

	/**
	 *
	 * @param fileName
	 * @return
	 */
	fun loadAudio(fileName: String): Int {
		val path = searchPaths.map {
			File(it, fileName)
		}.firstOrNull { it.exists() } ?: error("file $fileName not found in search paths in ${searchPaths.joinToString()}")

		// From file
		var stream = AudioSystem.getAudioInputStream(path)

		// At present, ALAW and ULAW encodings must be converted
		// to PCM_SIGNED before it can be played
		var format = stream.format
		if (format.encoding !== AudioFormat.Encoding.PCM_SIGNED) {
			format = AudioFormat(
				AudioFormat.Encoding.PCM_SIGNED,
				format.sampleRate,
				format.sampleSizeInBits * 2,
				format.channels,
				format.frameSize * 2,
				format.frameRate,
				true) // big endian
			stream = AudioSystem.getAudioInputStream(format, stream)
		}

		// Create the clip
		val info = DataLine.Info(
			Clip::class.java, stream.format, stream.frameLength.toInt() * format.frameSize)
		val clip = AudioSystem.getLine(info) as Clip

		// This method does not return until the audio file is completely loaded
		clip.open(stream)
		sounds.add(clip)
		log.info("Loaded $fileName -> ${sounds.size}")
		return sounds.size - 1
	}

	/**
	 *
	 * @param id
	 */
	fun playSound(id: Int, loops: Int = 0) {
		sounds.getOrNull(id)?.let {
			it.stop()
			it.framePosition = 0
			it.loop(loops)
			it.start()
		} ?: log.error("Invalid sound id $id")
	}

	private val sounds = ArrayList<Clip>(32)
}